pkg/controller/jobs/mpijob/mpijob_controller.go

```diff
@@ -125,11 +125,10 @@ func (j *MPIJob) RunWithNodeAffinity(nodeSelectors []map[string]string) {
 	}
 }
 
-func (j *MPIJob) RestoreNodeAffinity(podSets []kueue.PodSet) {
+func (j *MPIJob) RestoreNodeAffinity(nodeSelectors []map[string]string) {
 	orderedReplicaTypes := orderedReplicaTypes(&j.Spec)
-	for index := range podSets {
+	for index, nodeSelector := range nodeSelectors {
 		replicaType := orderedReplicaTypes[index]
```

Comment:
the serialization is assuming that podsets can be ordered. While this works for Job and MPIJob, I prefer not to make that assumption in general. Let's add the name of the podset to the serialization.

Reply:
The same approach is used to provide the pod sets to the workload. Sure it could be done but it will require the change of the source for original selectors (get them from them from the job not workload) ... Maybe we can create a follow-up issue for this.

Reply:
```golang
type PodSetNodeSelector struct {
  Name string
  NodeSelector map[string]string
}
```

Then you serialize/deserialize `[]PodSetNodeSelector`

Reply:
These kind of changes cannot be easily done in follow ups if we want to provide backwards compatibility guarantees.

Reply:
The name of the podset cold be change within a workload, by the defaulting web hook, there is no guarantee tht the name you get from the workload can be used in restore.

Reply:
if a defaulting webhook changes the name that's ok.

You should have the name defaulted at the time when the workload is admitted.

Reply:
But that means that you cannot (100% reliably ) use the name in the restore logic.